README.md
# Laravel 5 IsoCodes Validation

[![Latest Version on Packagist](https://img.shields.io/packagist/v/pixelpeter/laravel5-isocodes-validation.svg?style=flat-square&t=123)](https://packagist.org/packages/pixelpeter/laravel5-isocodes-validation)
[![Software License](https://img.shields.io/badge/license-GPL-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Travis Build](https://img.shields.io/travis/pixelpeter/laravel5-isocodes-validation/master.svg?style=flat-square)](https://travis-ci.org/pixelpeter/laravel5-isocodes-validation)
[![Scrutinizer Quality](https://img.shields.io/scrutinizer/g/pixelpeter/laravel5-isocodes-validation.svg?style=flat-square)](https://scrutinizer-ci.com/g/pixelpeter/laravel5-isocodes-validation)
[![Scrutinizer Build](https://img.shields.io/scrutinizer/build/g/pixelpeter/laravel5-isocodes-validation.svg?style=flat-square)](https://scrutinizer-ci.com/g/pixelpeter/laravel5-isocodes-validation)
[![SensioLabsInsight](https://img.shields.io/sensiolabs/i/32635b20-a120-46de-a1af-4ce876bdcfbe.svg?style=flat-square)](https://insight.sensiolabs.com/projects/32635b20-a120-46de-a1af-4ce876bdcfbe)
[![Total Downloads](https://img.shields.io/packagist/dt/pixelpeter/laravel5-isocodes-validation.svg?style=flat-square)](https://packagist.org/packages/pixelpeter/laravel5-isocodes-validation)
[![Coverage Status](https://coveralls.io/repos/github/pixelpeter/laravel5-isocodes-validation/badge.svg?branch=master)](https://coveralls.io/github/pixelpeter/laravel5-isocodes-validation?branch=master)

A simple Laravel 5 wrapper for the [IsoCodes Validation library](https://github.com/ronanguilloux/IsoCodes) from ronanguilloux.

## Installation

### Step 1: Install Through Composer
``` bash
composer require pixelpeter/laravel5-isocodes-validation
```

### Step 2: Add the Service Provider
*(not needed starting with v2.x because of auto discovery)*

Add the service provider in `app/config/app.php`
```php
'provider' => [
    ...
    Pixelpeter\IsoCodesValidation\IsoCodesValidationServiceProvider::class,
    ...
];
```

## Usage

### Simple examples

```php
// Checking out your e-commerce shopping cart?
$payload = [
    'creditcard' => '12345679123456'
];
$rules = [
    'creditcard' => 'creditcard'
];

$validator = Validator::make($payload, $rules);
```

### Examples with reference parameter
Some rules need a reference to be validated against (e.g. `country` for `zipcode`).

Just pass the name of the field holding the reference to the rule.

```php
// Sending letters to the Labrador Islands ?
$payload = [
    'zipcode' => 'A0A 1A0',
    'country' => 'CA'
];
$rules = [
    'zipcode' => 'zipcode:country'
];

$validator = Validator::make($payload, $rules);

// Publishing books?
$payload = [
    'isbn' => '2-2110-4199-X',
    'isbntype' => 13
];
$rules = [
    'zipcode' => 'isbn:isbntype'
];

$validator = Validator::make($payload, $rules);
```

### Example with arrays and dot notation
*(added in v3.x)*

As suggested by @otr-tomek I've added support for all validation methods using arrays in dot notation as an input.

```php
$payload = [
    'data' => [
        [
            'country' => 'DE',
            'zipcode' => 63741
        ],
        [
            'country' => 'AT',
            'zipcode' => 1180
        ]
  ] 
];

$validator = Validator::make($payload, [
    'data.*.zipcode' => 'zipcode:data.*.country'
]);
```

### Validation error messages
Error messages can contain the name and value of the field and the value of the reference
```php
$payload = [
    'phonenumber' => 'invalid',
    'country' => 'GB'
];
$rules = [
    'phonenumber' => 'phonenumber:country'
];

$validator = Validator::make($payload, $rules);

print $validator->errors()->first(); // The value "invalid" of phonenumber is not valid for "GB".
```

### More Examples
Refer to [IsoCodes Validation library](https://github.com/ronanguilloux/IsoCodes) for more examples and documentation.

## Testing
Run the tests with:
```bash
vendor/bin/phpunit
```

## License

GNU General Public License v3.0 only. Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All Notable changes for the Laravel 5 IsoCodes Validation  will be documented in this file

## 3.0.0
- Fix issue #8: missing function replaceVat
- Laravel 5.8
- Upgrade to phpunit 7.5.x
- Add php 7.3 to travis-ci
- Add support for arrays with dot notations to all validation methods
  ```php
  $payload = [
      'data' => [
          [
              'country' => 'DE',
              'zipcode' => 63741
          ],
          [
              'country' => 'AT',
              'zipcode' => 1180
          ]
    ] 
  ];
  
  $validator = Validator::make($payload, [
      'data.*.zipcode' => 'zipcode:data.*.country'
  ]);
  ```
- Refactor: fix scrutinizer issues  
- Refactor: add (external) fixtures to tests to reduce number of lines
- Refactor: remove unused function params

## 2.0.0
- Laravel 5.5 compatibility with Auto-Discovery
- Require php 7.0+
- Upgrade to phpunit 6.5.x
- Add php 7.2 to travis-ci

## 1.2.0
Added support for following validation rules
- BSN
- GDTI
- GLN
- GRAI
- GSRN
- ISMN
- ISWC
- UDI
- UPCA

## 1.1.0
- Laravel 5.4 compatibility 

## 1.0.0
- Stable first release


tests/TestCase.php
<?php

namespace Pixelpeter\IsoCodesValidation;

abstract class TestCase extends \Illuminate\Foundation\Testing\TestCase
{
    /**
     * The base URL to use while testing the application.
     *
     * @var string
     */
    protected $baseUrl = 'http://localhost';

    /**
     * Creates the application.
     *
     * @return \Illuminate\Foundation\Application
     */
    public function createApplication()
    {
        $app = require __DIR__.'/../vendor/laravel/laravel/bootstrap/app.php';

        $app->register(\Pixelpeter\IsoCodesValidation\IsoCodesValidationServiceProvider::class);

        $app->make(\Illuminate\Contracts\Console\Kernel::class)->bootstrap();

        return $app;
    }
}


tests/fixtures/valid.php
<?php

return [
    ['bban', '15459450000411700920U62'],
    ['bsn', '075278431'],
    ['cif', 'N0032484H'],
    ['creditcard', '4111111111111111'],
    ['ean8', '90311017'],
    ['ean13', '4719512002889'],
    ['gdti', '4719512002889 1234567890 123456'],
    ['gln', '0614141000012'],
    ['grai', '04719512002889 1234567890 12345A'],
    ['gsrn', '735005385000000011'],
    ['gtin8', '12345670'],
    ['gtin12', '1-23456-78999-9'],
    ['gtin13', '4006381333931'],
    ['gtin14', '12345678901231'],
    ['iban', 'AL86751639367318444714198669'],
    ['insee', '253012B073004'],
    ['ipaddress', '73.194.66.94'],
    ['isin', 'US0378331005'],
    ['ismn', '979-0-2600-0043-8'],
    ['iswc', 'T-000.000.001-0'],
    ['mac', '01-2d-4c-ef-89-59'],
    ['nif', '04381012H'],
    ['sedol', 'B0YBKJ7'],
    ['siren', 523311082],
    ['siret', 33493272000017],
    ['sscc', '340123450000000000'],
    ['ssn', '651-01-3431'],
    ['structured_communication', '123456789002'],
    ['swift_bic', 'BCEELULL'],
    ['udi', '07610221010301'],
    ['uknin', 'EH123456A'],
    ['upca', '1-23456-78999-9'],
    ['vat', 'LV12345678901'],
];


tests/fixtures/invalid.php
<?php

return [
    // BBAN
    [
        'payload' => [
            'bban' => 'invalid'
        ],
        'rules' => [
            'bban' => 'bban'
        ],
        'message' => 'The value "invalid" of bban is not a valid BBAN code.'
    ],
    // BSN
    [
        'payload' => [
            'bsn' => 'invalid'
        ],
        'rules' => [
            'bsn' => 'bsn'
        ],
        'message' => 'The value "invalid" of bsn is not a valid BSN.'
    ],
    // CIF
    [
        'payload' => [
            'cif' => 'invalid'
        ],
        'rules' => [
            'cif' => 'cif'
        ],
        'message' => 'The value "invalid" of cif is not a valid CIF code.'
    ],
    // CREDITCARD
    [
        'payload' => [
            'creditcard' => 'invalid'
        ],
        'rules' => [
            'creditcard' => 'creditcard'
        ],
        'message' => 'The value "invalid" of creditcard is not a valid credit card number.'
    ],
    // EAN8
    [
        'payload' => [
            'ean8' => 'invalid'
        ],
        'rules' => [
            'ean8' => 'ean8'
        ],
        'message' => 'The value "invalid" of ean8 is not a valid EAN8 code.'
    ],
    // EAN13
    [
        'payload' => [
            'ean13' => 'invalid'
        ],
        'rules' => [
            'ean13' => 'ean13'
        ],
        'message' => 'The value "invalid" of ean13 is not a valid EAN13 code.'
    ],
    // GDTI
    [
        'payload' => [
            'gdti' => 'invalid'
        ],
        'rules' => [
            'gdti' => 'gdti'
        ],
        'message' => 'The value "invalid" of gdti is not a valid Global Document Type Identifier (GDTI).'
    ],
    // GLN
    [
        'payload' => [
            'gln' => 'invalid'
        ],
        'rules' => [
            'gln' => 'gln'
        ],
        'message' => 'The value "invalid" of gln is not a valid Global Location Number (GLN).'
    ],
    // GRAI
    [
        'payload' => [
            'grai' => 'invalid'
        ],
        'rules' => [
            'grai' => 'grai'
        ],
        'message' => 'The value "invalid" of grai is not a valid Global Returnable Asset Identifier.'
    ],
    // GSRN
    [
        'payload' => [
            'gsrn' => 'invalid'
        ],
        'rules' => [
            'gsrn' => 'gsrn'
        ],
        'message' => 'The value "invalid" of gsrn is not a valid Global Service Relation Number (GS1).'
    ],
    // GTIN8
    [
        'payload' => [
            'gtin8' => 'invalid'
        ],
        'rules' => [
            'gtin8' => 'gtin8'
        ],
        'message' => 'The value "invalid" of gtin8 is not a valid GTIN-8 code.'
    ],
    // GTIN12
    [
        'payload' => [
            'gtin12' => 'invalid'
        ],
        'rules' => [
            'gtin12' => 'gtin12'
        ],
        'message' => 'The value "invalid" of gtin12 is not a valid GTIN-12 code.'
    ],
    // GTIN13
    [
        'payload' => [
            'gtin13' => 'invalid'
        ],
        'rules' => [
            'gtin13' => 'gtin13'
        ],
        'message' => 'The value "invalid" of gtin13 is not a valid GTIN-13 code.'
    ],
    // GTIN14
    [
        'payload' => [
            'gtin14' => 'invalid'
        ],
        'rules' => [
            'gtin14' => 'gtin14'
        ],
        'message' => 'The value "invalid" of gtin14 is not a valid GTIN-14 code.'
    ],
    // IBAN
    [
        'payload' => [
            'iban' => 'invalid'
        ],
        'rules' => [
            'iban' => 'iban'
        ],
        'message' => 'The value "invalid" of iban is not a valid IBAN.'
    ],
    // INSEE
    [
        'payload' => [
            'insee' => 'invalid'
        ],
        'rules' => [
            'insee' => 'insee'
        ],
        'message' => 'The value "invalid" of insee is not a valid INSEE code.'
    ],
    // IP
    [
        'payload' => [
            'ipaddress' => 'invalid'
        ],
        'rules' => [
            'ipaddress' => 'ipaddress'
        ],
        'message' => 'The value "invalid" of ipaddress is not a valid ip address.'
    ],
    // ISBN
    [
        'payload' => [
            'isbn' => 'invalid',
            'type' => 13
        ],
        'rules' => [
            'isbn' => 'isbn:type'
        ],
        'message' => 'The value "invalid" of isbn is not a valid ISBN.'
    ],
    // ISIN
    [
        'payload' => [
            'isin' => 'invalid'
        ],
        'rules' => [
            'isin' => 'isin'
        ],
        'message' => 'The value "invalid" of isin is not a valid ISIN.'
    ],
    // ISMN
    [
        'payload' => [
            'ismn' => 'invalid'
        ],
        'rules' => [
            'ismn' => 'ismn'
        ],
        'message' => 'The value "invalid" of ismn is not a valid International Standard Music Number or ISMN (ISO 10957)',
    ],
    // ISWC
    [
        'payload' => [
            'iswc' => 'invalid'
        ],
        'rules' => [
            'iswc' => 'iswc'
        ],
        'message' => 'The value "invalid" of iswc is not a valid International Standard Musical Work Code (ISWC)'
    ],
    // MAC
    [
        'payload' => [
            'mac' => 'invalid'
        ],
        'rules' => [
            'mac' => 'mac'
        ],
        'message' => 'The value "invalid" of mac is not a valid MAC address.'
    ],
    // NIF
    [
        'payload' => [
            'nif' => 'invalid'
        ],
        'rules' => [
            'nif' => 'nif'
        ],
        'message' => 'The value "invalid" of nif is not a valid NIF.'
    ],
    // ORGANISME_TYPE12_NORME_B2
    [
        'payload' => [
            'organisme_type12_norme_b2' => 'invalid',
            'clef' => 2
        ],
        'rules' => [
            'organisme_type12_norme_b2' => 'organisme_type12_norme_b2:clef'
        ],
        'message' => 'The value "invalid" of organisme type12 norme b2 is not a valid Organisme Type12 Norme B2.'
    ],
    // PHONE NUMBER
    [
        'payload' => [
            'country' => 'GB',
            'phonenumber' => 'invalid'
        ],
        'rules' => [
            'phonenumber' => 'phonenumber:country'
        ],
        'message' => 'The value "invalid" of phonenumber is not valid for "GB".'
    ],
    // SEDOL
    [
        'payload' => [
            'sedol' => 'invalid'
        ],
        'rules' => [
            'sedol' => 'sedol'
        ],
        'message' => 'The value "invalid" of sedol is not a valid SEDOL.'
    ],
    // SIREN
    [
        'payload' => [
            'siren' => 'invalid'
        ],
        'rules' => [
            'siren' => 'siren'
        ],
        'message' => 'The value "invalid" of siren is not a valid SIREN code.'
    ],
    // SIRET
    [
        'payload' => [
            'siret' => 'invalid'
        ],
        'rules' => [
            'siret' => 'siret'
        ],
        'message' => 'The value "invalid" of siret is not a valid SIRET code.'
    ],
    // SSCC
    [
        'payload' => [
            'sscc' => 'invalid'
        ],
        'rules' => [
            'sscc' => 'sscc'
        ],
        'message' => 'The value "invalid" of sscc is not a valid SSCC.'
    ],
    // SSN
    [
        'payload' => [
            'ssn' => 'invalid'
        ],
        'rules' => [
            'ssn' => 'ssn'
        ],
        'message' => 'The value "invalid" of ssn is not a valid SSN.'
    ],
    // STRUCTURED_COMMUNICATION
    [
        'payload' => [
            'structured_communication' => 'invalid'
        ],
        'rules' => [
            'structured_communication' => 'structured_communication'
        ],
        'message' => 'The value "invalid" of structured communication is not a valid structured communication code.'
    ],
    // SWIFT_BIC
    [
        'payload' => [
            'swift_bic' => 'invalid'
        ],
        'rules' => [
            'swift_bic' => 'swift_bic'
        ],
        'message' => 'The value "invalid" of swift bic is not a valid SWIFT/BIC.'
    ],
    // UDI
    [
        'payload' => [
            'udi' => 'invalid'
        ],
        'rules' => [
            'udi' => 'udi'
        ],
        'message' => 'The value "invalid" of udi is not a valid UDI.'
    ],
    // UKNIN
    [
        'payload' => [
            'uknin' => 'invalid'
        ],
        'rules' => [
            'uknin' => 'uknin'
        ],
        'message' => 'The value "invalid" of uknin is not a valid UK National Insurance Number.'
    ],
    // UPCA
    [
        'payload' => [
            'upca' => 'invalid'
        ],
        'rules' => [
            'upca' => 'upca'
        ],
        'message' => 'The value "invalid" of upca is not a valid UPCA.'
    ],
    // VAT
    [
        'payload' => [
            'vat' => 'invalid'
        ],
        'rules' => [
            'vat' => 'vat'
        ],
        'message' => 'The value "invalid" of vat is not a valid VAT.'
    ],
    // ZIPCODE
    [
        'payload' => [
            'country' => 'GB',
            'zipcode' => 63741
        ],
        'rules' => [
            'zipcode' => 'zipcode:country'
        ],
        'message' => 'The value "63741" of zipcode is not valid for "GB".'
    ],
];


src/IsoCodesValidationServiceProvider.php
<?php

namespace Pixelpeter\IsoCodesValidation;

use Illuminate\Support\ServiceProvider;

class IsoCodesValidationServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap the application services.
     *
     * @return void
     */
    public function boot()
    {
        // load translation files
        $this->loadTranslationsFrom(
            __DIR__ . '/../lang',
            'validation'
        );

        // registering intervention validator extension
        $this->app['validator']->resolver(function ($translator, $data, $rules, $messages, $customAttributes) {
            // set the validation error messages
            foreach (get_class_methods('Pixelpeter\IsoCodesValidation\IsoCodesValidator') as $method) {
                $key = $this->getTranslationKeyFromMethodName($method);

                $messages[$key] = $this->getErrorMessage($translator, $messages, $key);
            }

            return new IsoCodesValidator($translator, $data, $rules, $messages, $customAttributes);
        });
    }

    /**
     * Return translation key for correspondent method name
     *
     * @param  string $name
     * @return string
     */
    private function getTranslationKeyFromMethodName($name)
    {
        if (stripos($name, 'validate') !== false) {
            return /** @scrutinizer ignore-deprecated */ snake_case(substr($name, 8));
        }

    }

    /**
     * Return the matching error message for the key
     *
     * @param  string $key
     * @return string
     */
    private function getErrorMessage($translator, $messages, $key)
    {
        // return error messages passed directly to the validator
        if (isset($messages[$key])) {
            return $messages[$key];
        }

        // return error message from validation translation file
        if ($translator->has("validation.{$key}")) {
            return $translator->get("validation.{$key}");
        }

        // return packages default message
        return $translator->get("validation::validation.{$key}");
    }

    /**
     * Register the application services.
     *
     * @return void
     */
    public function register()
    {
        //
    }
}


lang/en/validation.php
<?php

return [
    'bban'                      => 'The value ":value" of :attribute is not a valid BBAN code.',
    'bsn'                       => 'The value ":value" of :attribute is not a valid BSN.',
    'cif'                       => 'The value ":value" of :attribute is not a valid CIF code.',
    'creditcard'                => 'The value ":value" of :attribute is not a valid credit card number.',
    'ean8'                      => 'The value ":value" of :attribute is not a valid EAN8 code.',
    'ean13'                     => 'The value ":value" of :attribute is not a valid EAN13 code.',
    'gdti'                      => 'The value ":value" of :attribute is not a valid Global Document Type Identifier (GDTI).',
    'gln'                       => 'The value ":value" of :attribute is not a valid Global Location Number (GLN).',
    'grai'                      => 'The value ":value" of :attribute is not a valid Global Returnable Asset Identifier.',
    'gsrn'                      => 'The value ":value" of :attribute is not a valid Global Service Relation Number (GS1).',
    'gtin8'                     => 'The value ":value" of :attribute is not a valid GTIN-8 code.',
    'gtin12'                    => 'The value ":value" of :attribute is not a valid GTIN-12 code.',
    'gtin13'                    => 'The value ":value" of :attribute is not a valid GTIN-13 code.',
    'gtin14'                    => 'The value ":value" of :attribute is not a valid GTIN-14 code.',
    'iban'                      => 'The value ":value" of :attribute is not a valid IBAN.',
    'insee'                     => 'The value ":value" of :attribute is not a valid INSEE code.',
    'ipaddress'                 => 'The value ":value" of :attribute is not a valid ip address.',
    'isbn'                      => 'The value ":value" of :attribute is not a valid ISBN.',
    'isin'                      => 'The value ":value" of :attribute is not a valid ISIN.',
    'ismn'                      => 'The value ":value" of :attribute is not a valid International Standard Music Number or ISMN (ISO 10957)',
    'iswc'                      => 'The value ":value" of :attribute is not a valid International Standard Musical Work Code (ISWC)',
    'mac'                       => 'The value ":value" of :attribute is not a valid MAC address.',
    'nif'                       => 'The value ":value" of :attribute is not a valid NIF.',
    'organisme_type12_norme_b2' => 'The value ":value" of :attribute is not a valid Organisme Type12 Norme B2.',
    'phonenumber'               => 'The value ":value" of :attribute is not valid for ":country".',
    'sedol'                     => 'The value ":value" of :attribute is not a valid SEDOL.',
    'siren'                     => 'The value ":value" of :attribute is not a valid SIREN code.',
    'siret'                     => 'The value ":value" of :attribute is not a valid SIRET code.',
    'sscc'                      => 'The value ":value" of :attribute is not a valid SSCC.',
    'ssn'                       => 'The value ":value" of :attribute is not a valid SSN.',
    'structured_communication'  => 'The value ":value" of :attribute is not a valid structured communication code.',
    'swift_bic'                 => 'The value ":value" of :attribute is not a valid SWIFT/BIC.',
    'udi'                       => 'The value ":value" of :attribute is not a valid UDI.',
    'uknin'                     => 'The value ":value" of :attribute is not a valid UK National Insurance Number.',
    'upca'                      => 'The value ":value" of :attribute is not a valid UPCA.',
    'vat'                       => 'The value ":value" of :attribute is not a valid VAT.',
    'zipcode'                   => 'The value ":value" of :attribute is not valid for ":country".',
];
